rlapy/comps/evd.py

```diff
@@ -0,0 +1,432 @@
+    nu = np.sqrt(n)*epsilon_mach*la.norm(Y)
+    # a temporary regularization parameter
+    Y = Y + nu*S
+    R = la.cholesky(S.T @ Y, lower=True)
+    # R is upper-triangular and R^T @ R = S^T @ Y = S^T @ (A + nu*I)S
+    B = Y @ la.inv(R.T)
+    # B has n rows and k + s columns
+    V, Sigma_matrix, Wh = la.svd(B)
+    W = Wh.T
+    
+    comp_list = [k]
+    for i in range(min(k,n)):
+        if Sigma_matrix[(i+1)]**2<=nu:
+            comp_list.append(i)
+    #comp_list constracuts the union from which we drop components next.        
+    r = min(comp_list) 
+    # drop components that relied on regularization
+    lambda_matrix = (Sigma_matrix**2)[range(r)]-nu
+    V = V[:,range(r)]
+    lambda_matrix = np.diag(lambda_matrix)
+    return V, lambda_matrix
+
+###############################################################################
+#       Object-oriented interfaces
+###############################################################################
+
+class EVDecomposer:
+
+    TOL_CONTROL = 'none'
+
+    def __call__(self, A, k, epsilon, s, rng):
+        """
+        Return a matrix V with orthonormal columns and a lambda matrix (in vector form) lambda_matrix where
+        the diagonal are the eigen values of 
+
+        Parameters
+        ----------
+        A : Union[ndarray, spmatrix, LinearOperator]
+            Data matrix to approximate. A must be an n × n Hermitian matrix. 
+
+        k : int
+            Target for the number of columns in Q: 0 < k <= min(A.shape).
+            Typically, k << min(A.shape). Conformant implementations ensure
+            Q has at most k columns. For certain implementations it's
+            reasonable to choose k as large as k = min(A.shape), in which
+            case the implementation returns only once a specified error
+            tolerance has been met.
+
+        epsilon : float
+            Target accuracy for the oversampled approximation of A: 0 < epsilon < np.inf.
+            This parameter inherits from the QBDecomposer or RangeFinder class.
+            
+        s  : int
+            Auxiliary parameter for the QBDecomposer or RangeFinder.
+
+        rng : Union[None, int, SeedSequence, BitGenerator, Generator]
+            Determines the numpy Generator object that manages any and all
+            randomness in this function call.
+
+        Returns
+        -------
+        V : ndarray
+            Has shape (A.shape[0], k). Columns are orthonormal.
+
+        lambda_matrix : ndarray
+            Has shape (k,k). lambda_matrix is a diagonal matrix storing the eigenvalues of 
+            the matrix A
+        """
+        raise NotImplementedError()
+
+
+class EVD1(EVDecomposer):
+
+    TOL_CONTROL = 'unknown'  # depends on implementation of rangefinder
+
+    def __init__(self, rf: RangeFinder):
+        """
+        Parameters
+        ----------
+        rf : RangeFinder
+        """
+        self.rangefinder = rf
```

Comment:
It's actually fine to use any QBDecomposer. The QBDecomposer interface requires that ``(Q, B)`` satisfy ``B = Q.T @ A`` (at least when the algorithm runs in exact arithmetic).